### DISCUSSION TITLE
Storage threat model tutorial

### DISCUSSION THREAD

**Discussion by BenWestgate** [BenWestgate/Bails#90] [Ideas]
From telegram, posted her for easier access later:

![image](https://github.com/BenWestgate/Bails/assets/73506583/b71f1cd2-7808-46a1-bc1f-6723fe1d2689)

Ben Westgate, [9/22/23 7:45 AM]
I'm thinking after Beta I'll try to have a graphical utility that walks users thru a threat model of their storage locations and persons.  Picks the correct threshold, share quantity and how many copies of each share to distribute and to whom after asking simple questions.

Ben Westgate, [9/22/23 7:46 AM]
It's arguably the most important part after creating a backup and it's strange that nobody has made a tutorial for it. As it is something a computer can optimize for a user.

@extrapockets extrapockets, [9/22/23 7:58 AM]
This is a great idea.

@epiccurious Epic Curious, [9/22/23 8:35 AM]
Agreed, it would be a nice touch.

**Comment by BenWestgate** [comment]
Incomplete Things this would need to do:
- Ask the user if they want to be able to recover their seed without asking other people (be a single point of failure ok for savings, not for inheritance)
- Ask user to list the secure locations the user can personally come up with and label them. (Give suggestions home safe, wallet, safe deposit box, locked drawer at work, memorized passphrase etc)
- Ask user to list the professionals they trust could secure shares for them (estate lawyer, accountant, etc)
- Ask user to list all their heirs (give suggestions: parents, wife, children, in some cases siblings, nephews, uncles, grandparents, grandchildren)
- Ask user which of these heirs they trust can secure a share or secret for them and which can not.
- Go thru the list of locations and people and rate them 1 to 3 by how likely your heirs will be able to access them. (hidden too well, uncooperative custodian, confiscated at death, betrayal etc.)
- Go thru the list of locations and people and rate them 1 to 3 by how likely they are to lose your data. (natural disaster, forgetfulness, maliciously withholding data, carelessness, memory loss, death, etc.)
- Go thru the list of locations and people and rate them 1 to 3 by how likely they data they hold is to fall into wrong hands. (burglary, betrayal, seizure, confiscation, carelessness, torture, judicial action etc.)
- Perform a Round Robin over every possible threshold set of share assignment combinations and rate the combos 1 to 3 how likely ALL 2 or 3 fall into wrong hands (same theft risks, collusion risk, rubber hose risks, ransom risks, blabbermouth risks, are in the same home, city, country, etc)
- Perform a Round Robin over the different N - T + 1 combinations of locations and people, where N ranges from T to all the trusted items and T ranges and rate the likelihood all are lost from 1 to 3.
  - This would catch situations where all holders of a share face the same natural disaster risks, or risk of dying simultaneously.

Now the software can optimize the threshold, share quantity, how many copies for each share and who gets which share(s) or secret to minimize the risk of all cause loss to you and/or your heirs given the available locations, their theft and loss risk attributes as well as correlated theft and loss risks.

To make rating 1 to 3 more meaningful they might be defined as a percentage 1 = 5% per year, 2 = 10% per year, 3 = 20% per year that way the software actually finds a mathematically optimal solution.

Finally, it would print recovery instructions for your heirs with the name of the holder and share header of the share you need to write in and a grid for the CompactQR. These instructions have instructions on the conditions they'll release the data to you (healthy, well, and free or only in person etc.)

https://medium.com/@HodlSafe/instructions-for-signatories-dc7243f0acca
These are the best instructions template for your custodians I've seen online.

What's good about a computer tutorial for this is it can do an exhaustive search over every combination, including giving the same share to multiple people and giving one or more persons multiple unique shares or even the secret itself.

As the number of locations/people grows beyond 5-10 there is a combinatoric explosion of possibilities that makes it unlikely a human doing this by hand would find the optimum allocations for their shares.

**Reply by extrapockets** [reply]
I especially like the 5% per year, 10% per year, etc rating of trusted people because I believe it serves a dual function apart from computing the best sharing scheme - It will help users to think about the long-term risks and how they compound. This guided setup would be valuable to many. Shamir's Secret Sharing Scheme Sherpa.

**Reply by BenWestgate** [reply]
Another way that's perhaps more intuitive is to visualize for theft risk (which is usually lower than loss risk) a rating of "1" for 1-of-1 is as safe against theft as rating "3" is at "3-of-3", so roughly "triple" as likely.

Or for loss, one share rated "1" is as likely to be there in 4 years or next decade as any one of three shares rated "3".

That's true this a separate project from Bails or bails-wallet as any multisig or SSS user can solve for the ideal backup based on their perceived risks. It could even be extended to allow SLIP39 sub-sharing but I fear it's already too complex but I may be wrong. If Codex32 adds subsharing I will definitely support it. 

Before I would support subsharing, I would support resharing. That is where a master seed is split multiple times making multiple backup sets that do not combine together. I also should in the original release consider the case where some locations get the codex32 secret directly stored and other locations get shares of that secret.

**Comment by BenWestgate** [comment]
Since this is based on social key recovery, I'll leave this link here and try to finish Christopher Allen's book if I get stuck implementing this optimization problem and need heuristics to reduce the search complexity.
https://github.com/WebOfTrustInfo/rwot8-barcelona/blob/master/topics-and-advance-readings/social-key-recovery.md
